VariantShredding.md

```diff
@@ -25,290 +25,318 @@
+Readers can assume that a value is not an object if `typed_value` is null and that `typed_value` field values are correct; that is, readers do not need to read the `value` column if `typed_value` fields satisfy the required fields.
 
-```
-[
-  {
-    "a": [1, 2, 3],
-    "b": 100
-  },
-  {
-    "a": [4, 5, 6],
-    "b": 200
-  }
-]
-```
+## Nesting
 
+The `typed_value` associated with any Variant `value` field can be any shredded type, as shown in the sections above.
 
-The corresponding Parquet schema with “a” and “b” as leaf types is:
+For example, the `event` object above may also shred sub-fields as object (`location`) or array (`tags`).
 
 ```
-optional group variant_col {
- required binary metadata;
- optional binary variant_value;
- optional group array (LIST) {
-  repeated group list {
-   optional group element {
-    optional binary variant_value;
-    optional group object {
-     optional group a {
-      optional binary variant_value;
-      optional group array (LIST) {
-       repeated group list {
-        optional group element {
-         optional int64 typed_value;
-         optional binary variant_value;
+optional group event (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional group typed_value {
+    required group event_type {
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+    required group location {
+      optional binary value;
+      optional group typed_value {
+        required group latitude {
+          optional binary value;
+          optional double typed_value;
+        }
+        required group longitude {
+          optional binary value;
+          optional double typed_value;
+        }
+      }
+    }
+    required group tags {
+      optional binary value;
+      optional group typed_value (LIST) {
+        repeated group list {
+          required group element {
+            optional binary value;
+            optional binary typed_value (STRING);
+          }
         }
-       }
       }
-     }
-     optional group b {
-      optional int64 typed_value;
-      optional binary variant_value;
-     }
     }
-   }
   }
- }
 }
 ```
 
-In the above example schema, if “a” is an array containing a mix of integer and non-integer values, the engine will shred individual elements appropriately into either `typed_value` or `variant_value`.
-If the top-level Variant is not an array (for example, an object), the engine cannot shred the value and it will store it in the top-level `variant_value`.
-Similarly, if "a" is not an array, it will be stored in the `variant_value` under "a".
+# Data Skipping
 
-Consider the following example:
+Statistics for `typed_value` columns can be used for file, row group, or page skipping when `value` is always null (missing).
```

Comment:
```suggestion
Min/Max statistics for `typed_value` columns may be used for row group and page skipping when `value` is always null (missing).  The statistics may also be used if statistics can be used to infer the type of all non-null values have an irrelevant type (e.g.  if all values are variant null (`00`) and a filter specifies equality to a integer).
```

Reply:
I don't think that we should make this more specific because anything further risks making assumptions about the semantics of the filter. The example on the dev list demonstrates this. This can't state that strings can be ignored for numeric filters because the filter may cast a string and result in a valid value.